### PR TITLE
Log(n)-approximation shortest superstring algorithm

### PR DESCRIPTION
**Pull request by Tokixix**
Approximate shortest superstring based on 
Towards a DNA Sequencing Theory (Learning a String)

### CODE REVIEW

File: text/MateuszTokarz/opis.tex [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+Dowód powyższego lematu jest trywialny.
+\newline
+
+\textbf{Algorytm}
+\begin{enumerate}
+    \item Input to $S = \{s_1, \dots, s_n\}$. Jeżeli $|S| = 1$, to zwróć $s_1$.
+    \item Zdefiniujmy $T = \emptyset$.
+    \item Niech $s, s'$ będą słowami minimalizującymi 
+    $$\min_{m - \text{scalenie}} \frac{|m(s, s')|}{v(m(s, s'))},$$
```
Comment (krzysztof-turowski) [inline text/MateuszTokarz/opis.tex:47]: I guess it'd be less problematic if you'd define e.g. M = {m(s, s'): s, s' \in S} or something like that.
Still more formal it'd be if you define m(s, s', l) and assume that s has to be before s' and length of the overlap is l - then this is a proper function for some subset of (S, S, N).

However, the overall the idea is clear enough, just the formalities are not there.

File: text/MateuszTokarz/opis.tex [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+$$Cost(G_j) = \sum_{h=1}^{k} \sum_{s \in (G_j^{h+1} \backslash G_j^{h})} cost(s) = \sum_{h=1}^{k} \frac{|M_h(B_h, T_h))|}{v(M_h(B_h, T_h))} \cdot (||G_j^{h+1}|| - ||G_j^{h}||).$$
+Jako że nasz algorytm wybierał scalenia $M_h(B_h, T_h)$ realizujące minimalne wartości, to
+$$\sum_{h=1}^{k} \frac{|M_h(B_h, T_h))|}{v(M_h(B_h, T_h))} \cdot (||G_j^{h+1}|| - ||G_j^{h}||) \leq \sum_{h=1}^{k} \frac{|m_j(b_j, t_j))|}{v(m_j(b_j, t_j))} \cdot (||G_j^{h+1}|| - ||G_j^{h}||) = $$
+$$\sum_{h=1}^{k} \frac{|m_j(b_j, t_j))|}{||G_j^{h}||} \cdot (||G_j^{h+1}|| - ||G_j^{h}||) \leq |m_j(b_j, t_j))| \cdot Harm(||G_j||).$$
+Jako że $n$ jest co najwyżej wielomianowo większe niż każde $||G_j||$, to dostajemy oszacowanie
+$$Cost(G_j) \leq |m_j(b_j, t_j))| \cdot log(n),$$
+co z poprzednim lematem skutkuje w
+$$COST \leq 2nlog(n).$$
+
```
Comment (krzysztof-turowski) [inline text/MateuszTokarz/opis.tex:98]: Just one small question: can you think of an instance (in fact, some infinite family of instances for every n) which makes this bound tight?

File: approx_shortest_superstring/approximate_shortest_superstring.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  word_to_super = {word: list(filter(lambda merge_word : word[1:] in \
+  merge_word, all_merged_words)) for word in words}
+
+  merged_to_weight = get_merged_to_weight(all_merged_words, word_to_super)
+
+  result = []
+  current_words = set(words)
+  while len(current_words) > 1:
+    merged_candidate = None
+    candidate_weight = 1e9
```
Comment (krzysztof-turowski) [inline approx_shortest_superstring/approximate_shortest_superstring.py:65]: merged_candidate, candidate_weight = None, 10e9

File: approx_shortest_superstring/approximate_shortest_superstring.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    for (merged_word, weight_denominator) in merged_to_weight.items():
+      if (len(merged_word) / weight_denominator) < candidate_weight:
+        merged_candidate = merged_word
```
Comment (krzysztof-turowski) [inline approx_shortest_superstring/approximate_shortest_superstring.py:68]: Either I'm having a blackout or you use candidate_weight as constant (why 10e9? I suppose math.inf might be handy if you need infinity) - but aren't you supposed to pick the one with the smallest len(merged_word) / weight_denominator so don't you need to add "candidate_weight = len(merged_word) / weight_denominator" there?

File: approx_shortest_superstring/approximate_shortest_superstring.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    for word in words_to_remove:
+      current_words.remove(word)
+
+  for word in current_words:
+    result.append(word)
+
+  if len(result) == 1:
+    return result[0]
+
+  return shortest_common_super_approx(result)
```
Comment (krzysztof-turowski) [inline approx_shortest_superstring/approximate_shortest_superstring.py:87]: Well, python doesn't support tail-call optimization (at least without using some tricks) so please replace it with a large "while len(result) > 1:" loop (of course with appropriate modifications to the beginning of the function).

File: test/test_approximate_shortest_superstring.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+import random
+
+from approx_shortest_superstring.approximate_shortest_superstring \
+import shortest_common_super_approx
+
+from approx_shortest_superstring.exact_bruteforce \
+import shortest_common_super
+from generator import rand
+
+class TestCommentzWalter(unittest.TestCase):
```
Comment (krzysztof-turowski) [inline test/test_approximate_shortest_superstring.py:13]: TestShortestSuperstring

File: test/test_approximate_shortest_superstring.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
+import unittest
+import os
+import math
+import random
+
+from approx_shortest_superstring.approximate_shortest_superstring \
```
Comment (krzysztof-turowski) [inline test/test_approximate_shortest_superstring.py:6]: Nit: when using this type of linebreaks please add some spaces on the second line - it does not change anything with automatic styleguides, but I guess it's more readable than when both start equally.

File: text/MateuszTokarz/opis.tex [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+\newtheorem*{df}{Definicja}
+\newtheorem*{lemat}{Lemat}
+
+\begin{document}
+\title{Aproksymacja najkrótszego nadsłowa}
+\author{Mateusz Tokarz}
+\maketitle
+
+\section*{Opis Problemu}
+Problem najkrótszego nadsłowa w wersji decyzyjnej jest problemem NP zupełnym. W związku z tym pojawiało się wiele podejść do aproksymacji optymalizacyjnej wersji problemu. Tarhio i Ukkonen oraz Turner napisali dwie niezależne prace naukowe, w których opisano algorytmy aproksymacyjne dla rozważanego problemu. Mimo, że autorzy twierdzili, że skonstruowane przez nich algorytmy powinny być $2$-aproksymacyjne, to nie udało im się udowodnić żadnego ograniczenia.
```
Comment (krzysztof-turowski) [inline text/MateuszTokarz/opis.tex:21]: Oh, that's nice that you invoked Tarhio i Ukkonen and Turner works - can you provide the article names for them?

File: text/MateuszTokarz/opis.tex [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+Rozważmy teraz grupę $G_j$ i koszt, jaki płacimy za wyrzucenie jej elementów. Grupę $G_j$ dokładnie przed fazą $h$ będziemy oznaczać $G_j^{h}$.
+Sumę długości słów w zbiorze $A$ oznaczamy $||A||$.
+ 
+
+$$Cost(G_j) = \sum_{h=1}^{k} \sum_{s \in (G_j^{h+1} \backslash G_j^{h})} cost(s) = \sum_{h=1}^{k} \frac{|M_h(B_h, T_h))|}{v(M_h(B_h, T_h))} \cdot (||G_j^{h+1}|| - ||G_j^{h}||).$$
+Jako że nasz algorytm wybierał scalenia $M_h(B_h, T_h)$ realizujące minimalne wartości, to
+$$\sum_{h=1}^{k} \frac{|M_h(B_h, T_h))|}{v(M_h(B_h, T_h))} \cdot (||G_j^{h+1}|| - ||G_j^{h}||) \leq \sum_{h=1}^{k} \frac{|m_j(b_j, t_j))|}{v(m_j(b_j, t_j))} \cdot (||G_j^{h+1}|| - ||G_j^{h}||) = $$
+$$\sum_{h=1}^{k} \frac{|m_j(b_j, t_j))|}{||G_j^{h}||} \cdot (||G_j^{h+1}|| - ||G_j^{h}||) \leq |m_j(b_j, t_j))| \cdot Harm(||G_j||).$$
+Jako że $n$ jest co najwyżej wielomianowo większe niż każde $||G_j||$, to dostajemy oszacowanie
+$$Cost(G_j) \leq |m_j(b_j, t_j))| \cdot log(n),$$
```
Comment (krzysztof-turowski) [inline text/MateuszTokarz/opis.tex:95]: Use \log instead of log everywhere.

File: text/MateuszTokarz/opis.tex [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+\end{proof}
+
+Rozważmy teraz grupę $G_j$ i koszt, jaki płacimy za wyrzucenie jej elementów. Grupę $G_j$ dokładnie przed fazą $h$ będziemy oznaczać $G_j^{h}$.
+Sumę długości słów w zbiorze $A$ oznaczamy $||A||$.
+ 
+
+$$Cost(G_j) = \sum_{h=1}^{k} \sum_{s \in (G_j^{h+1} \backslash G_j^{h})} cost(s) = \sum_{h=1}^{k} \frac{|M_h(B_h, T_h))|}{v(M_h(B_h, T_h))} \cdot (||G_j^{h+1}|| - ||G_j^{h}||).$$
+Jako że nasz algorytm wybierał scalenia $M_h(B_h, T_h)$ realizujące minimalne wartości, to
+$$\sum_{h=1}^{k} \frac{|M_h(B_h, T_h))|}{v(M_h(B_h, T_h))} \cdot (||G_j^{h+1}|| - ||G_j^{h}||) \leq \sum_{h=1}^{k} \frac{|m_j(b_j, t_j))|}{v(m_j(b_j, t_j))} \cdot (||G_j^{h+1}|| - ||G_j^{h}||) = $$
+$$\sum_{h=1}^{k} \frac{|m_j(b_j, t_j))|}{||G_j^{h}||} \cdot (||G_j^{h+1}|| - ||G_j^{h}||) \leq |m_j(b_j, t_j))| \cdot Harm(||G_j||).$$
```
Comment (krzysztof-turowski) [inline text/MateuszTokarz/opis.tex:93]: What is Harm? I guess it's ||G_j||-th harmonic number, but frankly that's the first time I see such a notation.
Moreover, it's not entirely clear why sum_{h=1}^{k} \frac{||G_j^{h+1}|| - ||G_j^{h}||}{||G_j^{h}||} \le Harm(||G_j||) - please provide some one/two-line explanation.

**Comment by Tokixix** [comment]
> Overall nice and clean work. I have found only two several major issues:
> 
> * L68 in approximate_shortest_superstring.py - you don't update candidate_weight variable,
> * L93 in tex file - you don't explain how the sum is bounded by harmonic number.
> 
> Now I already put the grade in the spreadsheet, please let me know whether you want to finish this PR or leave it to me - I'd appreciate help, but I understand if you have some more important or urgent work.

I will finish the PR, thanks for the comments!

File: text/MateuszTokarz/opis.tex (hunk 2 above)
Comment (Tokixix) [reply text/MateuszTokarz/opis.tex:164]: In the version of the paper that I'd found at the beginning there was a belief that this is actually a 2-approximation algorithm. I couldn't think about any counter example, but then I found a newer version of the paper in which autor introduced a family which proves that the bound is tight. I added it to the description.

File: test/test_approximate_shortest_superstring.py (hunk 7 above)
Comment (Tokixix) [reply test/test_approximate_shortest_superstring.py:6]: I don't understand, can You please give me an example of such solution?

Comment (krzysztof-turowski) [reply test/test_approximate_shortest_superstring.py:6]: I was thinking of something like:
from something_very_very_long \
    import something

Don't worry, I'll correct it on my own later during merges of all your codes.

**Pull request merged**